optuna/_gp/gp.py

```diff
@@ -177,25 +180,33 @@ def loss_func(raw_params: np.ndarray) -> tuple[float, np.ndarray]:
             kernel_scale=torch.exp(raw_params_tensor[n_params]),
             noise_var=torch.exp(raw_params_tensor[n_params + 1]) + minimum_noise,
         )
+        if deterministic_objective:
+            params.noise_var = torch.tensor(minimum_noise, dtype=torch.float64)
         loss = -marginal_log_likelihood(
             torch.from_numpy(X), torch.from_numpy(Y), torch.from_numpy(is_categorical), params
         ) - log_prior(params)
         loss.backward()  # type: ignore
+        # scipy.minimize requires all the gradients to be zero for termination.
+        assert not deterministic_objective or raw_params_tensor.grad[n_params + 1] == 0
         return loss.item(), raw_params_tensor.grad.detach().numpy()  # type: ignore
 
     # jac=True means loss_func returns the gradient for gradient descent.
-    res = so.minimize(loss_func, initial_raw_params, jac=True)
+    res = so.minimize(
+        loss_func, initial_raw_params, jac=True, method="l-bfgs-b", options={"gtol": gtol}
+    )
     if not res.success:
         raise RuntimeError(f"Optimization failed: {res.message}")
 
-    # TODO(contramundum53): Handle the case where the optimization fails.
     raw_params_opt_tensor = torch.from_numpy(res.x)
 
-    return KernelParamsTensor(
+    res = KernelParamsTensor(
         inverse_squared_lengthscales=torch.exp(raw_params_opt_tensor[:n_params]),
         kernel_scale=torch.exp(raw_params_opt_tensor[n_params]),
         noise_var=torch.exp(raw_params_opt_tensor[n_params + 1]) + minimum_noise,
     )
+    if deterministic_objective:
+        res.noise_var = torch.tensor(minimum_noise, dtype=torch.float64)
```

Comment:
Can we do something like this so that we can keep frozen=True for KernelParamsTensor?
```suggestion
    noise_var = (
         torch.tensor(minimum_noise, dtype=torch.float64) if deterministic_objective
         else torch.exp(raw_params_opt_tensor[n_params + 1]) + minimum_noise
    )
    res = KernelParamsTensor(
        inverse_squared_lengthscales=torch.exp(raw_params_opt_tensor[:n_params]),
        kernel_scale=torch.exp(raw_params_opt_tensor[n_params]),
        noise_var=noise_var,
    )
```

Reply:
Personally keeping the `KernelParamsTensor` frozen doesn't particularly make sense, especially because what it has are `torch.Tensor`s whose gradients are designed to be rewritten. But I do not have a strong opinion.

Reply:
Done.

Reply:
This comment was made by mistake, but so i will count on you if you would like to revert the change.